src/lib/core/components/visualizations/StandardVisualization.tsx

```diff
@@ -0,0 +1,221 @@
+import Banner from '@veupathdb/coreui/dist/components/banners/Banner';
+import { getOrElse } from 'fp-ts/lib/Either';
+import { pipe } from 'fp-ts/lib/function';
+import { Decoder } from 'io-ts';
+import {
+  ComponentType,
+  useCallback,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react';
+import { useStudyEntities } from '../../hooks/study';
+import { useStudyMetadata } from '../../hooks/workspace';
+import { VariablesByInputName } from '../../utils/data-element-constraints';
+import { CustomSectionSpec, InputSpec, InputVariables } from './InputVariables';
+import { VisualizationProps } from './VisualizationTypes';
+
+type Config = Record<string, unknown>;
+
+export interface StandardPlotProps<T, U> {
+  vizConfig: U;
+  updateVizConfig: (config: U) => void;
+  data?: T;
+}
+
+interface StandardVsualizationProps<T, U> {
+  /**
+   * Describes what inputs the user can configure.
+   */
+  inputs: InputSpec[];
+  /**
+   * Determine if the showMissingness toggle should be shown, based on data
+   */
+  enableShowMissingnessToggle?: (data: T) => boolean;
+  /**
+   * Custom input selectors
+   */
+  customInputSections?: CustomSectionSpec[];
+  /**
+   * `io-ts` configuration `Decoder`
+   */
+  configDecoder: Decoder<unknown, U>;
+  /**
+   * Default configuration object, when one is not defined
+   */
+  defaultConfig: U;
+  /**
+   * A function that returns the ID of the output entity for the visualization
+   */
+  outputEntitySelector: (
+    selectedVariables: VariablesByInputName
+  ) => string | undefined;
+  /**
+   * Map selected variables to a configuration object
+   */
+  mapSelectedVariablesToConfig: (
+    selectedVariables: VariablesByInputName,
+    config: U
+  ) => U;
+  /**
+   * Map configuration object to selected variables
+   */
+  mapConfigToSelectedVariables: (config: U) => VariablesByInputName;
+  /**
+   * Map selected variables to a configuration object
+   */
+  mapShowMissingnessToConfig?: (showMissingness: boolean, config: U) => U;
+  /**
+   * Map configuration object to selected variables
+   */
+  mapConfigToShowMissingnes?: (config: U) => boolean;
+  /**
+   * A function that take a vizConfig object and returns a Promise for data.
+   */
+  getData: (vizConfig: U) => Promise<T>;
```

Comment:
I'm guessing you'll want `getData` to receive the data client as an argument?

Reply:
That's a good possibility. I don't really see a downside. Thanks!